mypy/checkexpr.py

```diff
@@ -2520,7 +2520,13 @@ def nonliteral_tuple_index_helper(self, left_type: TupleType, index: Expression)
                                       'actual type', 'expected type'):
             return AnyType(TypeOfAny.from_error)
         else:
-            return UnionType.make_simplified_union(left_type.items)
+            if isinstance(index, SliceExpr):
+                # if we index (1,2,3)[:name], we don't know what the return type is,
+                # beyond that it is a tuple of some sort.
+                anys = (AnyType(TypeOfAny.implementation_artifact),)
```

Comment:
I wasn't sure which variant to use here for TypeOfAny, so guidance would be appreciated.

Reply:
hmmm, I was thinking this would end up being `Tuple[Union[the tuple types], ...]` -- that way for homogenous tuples you'd end up with a consistent type

Reply:
or maybe only that specialization for homogenous tuples? 🤔

Reply:
Yeah, I was thinking that too, but then I thought about code like:
```
a = 2
s = (1, 3.3, '')[:a]
s[0] + s[1]
```
Though perhaps I am being too cautious.

Reply:
I could also special case homogeneous tuples to return `Tuple[T, ...]`, and for everything else fall back to `Tuple[Any, ...]`. Though for a decision like this I will defer to @JukkaL or someone else who wants to chime in here.

Reply:
that sounds fine to me 👍